### ISSUE TITLE
ClassCastException creating ORecordBytes field in distributed mode

### ISSUE THREAD

**Issue by phleemac** [orientechnologies/orientdb#4069]
I use ORecordBytes (http://orientdb.com/docs/last/Binary-Data.html "Store it with ORecordBytes" approach") to store image using GraphAPI.  There is no problem when running single node (start with bin/server.sh).   But it fails when run on 2 nodes in distribute mode, with exception:

```
2015-05-04 00:37:26:911 WARNING [dev-db-01] detected 1 node(s) in timeout or in conflict and quorum (2) has not been reached, rolling back changes for request (id=148 from=dev-
db-01 task=tx userName=admin) [ODistributedResponseManager]
2015-05-04 00:37:26:911 WARNING [dev-db-01] Quorum 2 not reached for request (id=148 from=dev-db-01 task=tx userName=admin). Timeout=22ms Servers in timeout/conflict are:
 - dev-db-02: java.lang.ClassCastException: com.orientechnologies.orient.core.record.impl.ORecordBytes cannot be cast to com.orientechnologies.orient.core.record.impl.ODocument
Received: {dev-db-02=java.lang.ClassCastException: com.orientechnologies.orient.core.record.impl.ORecordBytes cannot be cast to com.orientechnologies.orient.core.record.impl.OD
ocument, dev-db-01=java.lang.ClassCastException: com.orientechnologies.orient.core.record.impl.ORecordBytes cannot be cast to com.orientechnologies.orient.core.record.impl.ODocument} [ODistributedResponseManager]
2015-05-04 00:37:26:911 WARNING [dev-db-01] sending undo message for request (id=148 from=dev-db-01 task=tx userName=admin) to server dev-db-02 [ODistributedResponseManager]
2015-05-04 00:37:26:914 WARNING [dev-db-01] sending undo message for request (id=148 from=dev-db-01 task=tx userName=admin) to server dev-db-01 [ODistributedResponseManager]Sent run-time exception to the client /10.240.218.43:47281: com.orientechnologies.orient.server.distributed.ODistributedException: Error on executing distributed request (id=148 from=dev-db-01 task=tx userName=admin) against database 'massope_dev.[default, taglist_dev-db-01]' to nodes [dev-db-02, dev-db-01]
com.orientechnologies.orient.server.distributed.ODistributedException: Error on executing distributed request (id=148 from=dev-db-01 task=tx userName=admin) against database 'massope_dev.[default, taglist_dev-db-01]' to nodes [dev-db-02, dev-db-01]
        at com.orientechnologies.orient.server.hazelcast.OHazelcastDistributedDatabase.send2Nodes(OHazelcastDistributedDatabase.java:194)
        at com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin.sendRequest(OHazelcastPlugin.java:365)
        at com.orientechnologies.orient.server.distributed.ODistributedStorage.commit(ODistributedStorage.java:902)
        at com.orientechnologies.orient.core.tx.OTransactionOptimistic.doCommit(OTransactionOptimistic.java:483)
        at com.orientechnologies.orient.core.tx.OTransactionOptimistic.commit(OTransactionOptimistic.java:147)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2406)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2376)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.commit(ONetworkProtocolBinary.java:1082)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.executeRequest(ONetworkProtocolBinary.java:384)
        at com.orientechnologies.orient.server.network.protocol.binary.OBinaryNetworkProtocolAbstract.execute(OBinaryNetworkProtocolAbstract.java:217)
        at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:69)
Caused by: com.orientechnologies.orient.server.distributed.ODistributedException: Quorum 2 not reached for request (id=148 from=dev-db-01 task=tx userName=admin). Timeout=22ms Servers in timeout/conflict are:
 - dev-db-02: java.lang.ClassCastException: com.orientechnologies.orient.core.record.impl.ORecordBytes cannot be cast to com.orientechnologies.orient.core.record.impl.ODocument
Received: {dev-db-02=java.lang.ClassCastException: com.orientechnologies.orient.core.record.impl.ORecordBytes cannot be cast to com.orientechnologies.orient.core.record.impl.ODocument, dev-db-01=java.lang.ClassCastException: com.orientechnologies.orient.core.record.impl.ORecordBytes cannot be cast to com.orientechnologies.orient.core.record.impl.ODocument}
        at com.orientechnologies.orient.server.distributed.ODistributedResponseManager.manageConflicts(ODistributedResponseManager.java:576)
        at com.orientechnologies.orient.server.distributed.ODistributedResponseManager.getFinalResponse(ODistributedResponseManager.java:340)
        at com.orientechnologies.orient.server.hazelcast.OHazelcastDistributedDatabase.waitForResponse(OHazelcastDistributedDatabase.java:364)
        at com.orientechnologies.orient.server.hazelcast.OHazelcastDistributedDatabase.send2Nodes(OHazelcastDistributedDatabase.java:191)
        ... 10 more

```
 
I'm using version `2.0.8`

**Comment by lukaszkaleta** [comment]
Hi, could please show code snipped how you store/read files ?

**Comment by phleemac** [comment]
In scala:
```
...
          val dataRec = new ORecordBytes(graph.getRawGraph, props.data)
          dataRec.save()
          val vertex = graph.addVertex("class:Photo", 
            "contentType", props.contentType,
            "data", dataRec
          )
...
```
where `graph` is `OrientGraph`, `props.data` is byte array

**Comment by wouterv** [comment]
We have the exact same problem..

**Comment by wouterv** [comment]
The actual error is on a second node:

```
java.lang.ClassCastException: com.orientechnologies.orient.core.record.impl.ORecordBytes cannot be cast to com.orientechnologies.orient.core.record.impl.ODocument
	at com.orientechnologies.orient.server.distributed.task.OTxTask.execute(OTxTask.java:100)
	at com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin.executeOnLocalNode(OHazelcastPlugin.java:753)
	at com.orientechnologies.orient.server.hazelcast.ODistributedWorker.onMessage(ODistributedWorker.java:298)
	at com.orientechnologies.orient.server.hazelcast.ODistributedWorker.run(ODistributedWorker.java:121)
```

**Comment by wouterv** [comment]
Seems to me that this part:

```
          if (record != null)
            for (String f : ((ODocument) record).fieldNames()) {
              final Object fValue = ((ODocument) record).field(f);
              if (fValue instanceof ORecordLazyMultiValue)
                // DESERIALIZE IT TO ASSURE TEMPORARY RIDS ARE TREATED CORRECTLY
                ((ORecordLazyMultiValue) fValue).convertLinks2Records();
              else if (fValue instanceof ORecordId)
                ((ODocument) record).field(f, ((ORecordId) fValue).getRecord());
            }
```
should not do anything if the record is not of instance ODocument. An ORecordBytes does not have any fields, so this piece of code should not be executed.
So the if statement should read:
```
if (record instanceof ODocument)
```

Am I right?

**Comment by wouterv** [comment]
I tried the fix locally and it seems that it is working for us with the change above, will create a pull request.

**Comment by wouterv** [comment]
Can this be closed and released in 2.0.16?

**Comment by lvca** [comment]
We already merged it, so I'm closing it.
